evalml/model_understanding/graphs.py

```diff
@@ -622,7 +635,10 @@ def partial_dependence(pipeline, X, features, percentiles=(0.05, 0.95), grid_res
         if not (all([isinstance(x, str) for x in features]) or all([isinstance(x, int) for x in features])):
             raise ValueError("Features provided must be a tuple entirely of integers or strings, not a mixture of both.")
         is_categorical = [_is_feature_categorical(f, X) for f in features]
+        is_datetime = [_is_feature_datetime(f, X) for f in features]
         feature_names = _get_feature_names_from_str_or_col_index(X, features)
+        if any(is_datetime):
```

Comment:
Datetime columns run into the same problem as categorical columns where the sklearn pd implementation changes the dtypes to the first feature. However, unlike the categorical case we cannot solve the issue by moving the dt column first as datetime and numerical dtypes do not mix. Heres an example with just a date time featurizer and a logistic regression classifier.

```evalml/model_understanding/graphs.py:665: in partial_dependence
    avg_pred, values = sk_partial_dependence(wrapped, X=X, features=features, percentiles=percentiles, grid_resolution=grid_resolution)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/utils/validation.py:63: in inner_f
    return f(*args, **kwargs)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py:486: in partial_dependence
    estimator, grid, features_indices, X, response_method
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py:162: in _partial_dependence_brute
    pred = prediction_method(X_eval)
evalml/pipelines/components/utils.py:172: in predict_proba
    return _convert_woodwork_types_wrapper(self.pipeline.predict_proba(X).to_dataframe()).to_numpy()
evalml/pipelines/pipeline_meta.py:23: in _check_for_fit
    return method(self, X)
evalml/pipelines/binary_classification_pipeline.py:46: in predict_proba
    return super().predict_proba(X)
evalml/pipelines/pipeline_meta.py:23: in _check_for_fit
    return method(self, X)
evalml/pipelines/classification_pipeline.py:104: in predict_proba
    proba = self.estimator.predict_proba(X).to_dataframe()
evalml/pipelines/components/component_base_meta.py:27: in _check_for_fit
    return method(self, X)
evalml/pipelines/components/estimators/estimator.py:79: in predict_proba
    pred_proba = self._component_obj.predict_proba(X)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:1469: in predict_proba
    return super()._predict_proba_lr(X)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/linear_model/_base.py:323: in _predict_proba_lr
    prob = self.decision_function(X)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/linear_model/_base.py:284: in decision_function
    X = check_array(X, accept_sparse='csr')
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/utils/validation.py:63: in inner_f
    return f(*args, **kwargs)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/utils/validation.py:540: in check_array
    dtype_orig = np.result_type(*dtypes_orig)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (dtype('<M8[ns]'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), ...), kwargs = {}
relevant_args = (dtype('<M8[ns]'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), ...)

```
evalml/model_understanding/graphs.py:665: in partial_dependence
    avg_pred, values = sk_partial_dependence(wrapped, X=X, features=features, percentiles=percentiles, grid_resolution=grid_resolution)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/utils/validation.py:63: in inner_f
    return f(*args, **kwargs)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py:486: in partial_dependence
    estimator, grid, features_indices, X, response_method
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py:162: in _partial_dependence_brute
    pred = prediction_method(X_eval)
evalml/pipelines/components/utils.py:172: in predict_proba
    return _convert_woodwork_types_wrapper(self.pipeline.predict_proba(X).to_dataframe()).to_numpy()
evalml/pipelines/pipeline_meta.py:23: in _check_for_fit
    return method(self, X)
evalml/pipelines/binary_classification_pipeline.py:46: in predict_proba
    return super().predict_proba(X)
evalml/pipelines/pipeline_meta.py:23: in _check_for_fit
    return method(self, X)
evalml/pipelines/classification_pipeline.py:104: in predict_proba
    proba = self.estimator.predict_proba(X).to_dataframe()
evalml/pipelines/components/component_base_meta.py:27: in _check_for_fit
    return method(self, X)
evalml/pipelines/components/estimators/estimator.py:79: in predict_proba
    pred_proba = self._component_obj.predict_proba(X)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:1469: in predict_proba
    return super()._predict_proba_lr(X)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/linear_model/_base.py:323: in _predict_proba_lr
    prob = self.decision_function(X)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/linear_model/_base.py:284: in decision_function
    X = check_array(X, accept_sparse='csr')
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/utils/validation.py:63: in inner_f
    return f(*args, **kwargs)
../../.pyenv/versions/evalml/lib/python3.7/site-packages/sklearn/utils/validation.py:540: in check_array
    dtype_orig = np.result_type(*dtypes_orig)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (dtype('<M8[ns]'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), ...), kwargs = {}
relevant_args = (dtype('<M8[ns]'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'), ...)

>   ???
E   TypeError: The DTypes <class 'numpy.dtype[float64]'> and <class 'numpy.dtype[datetime64]'> do not have a common DType. For example they cannot be stored in a single array unless the dtype is `object`.
```

Reply:
@jeremyliweishih talked about this and converting to `object` dtype doesn't work either.